File: output/typescript/tests/07_params.ts
class WithParamsFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
    }
}


class WithParamsFrameContext {
    public event: WithParamsFrameEvent;
    public _return: any;
    public _data: Record<string, any>;

    constructor(event: WithParamsFrameEvent, default_return: any) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


class WithParamsCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: WithParamsCompartment | null;

    constructor(state: string, parent_compartment: WithParamsCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): WithParamsCompartment {
        const c = new WithParamsCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class WithParams {
    private _state_stack: Array<any>;
    private __compartment: WithParamsCompartment;
    private __next_compartment: WithParamsCompartment | null;
    private _context_stack: Array<any>;
    private total: number = 0;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.total = 0;
        this.__compartment = new WithParamsCompartment("Idle");
        this.__next_compartment = null;
        const __frame_event = new WithParamsFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: WithParamsFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new WithParamsFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new WithParamsFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new WithParamsFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: WithParamsFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: WithParamsCompartment) {
        this.__next_compartment = next_compartment;
    }

    public start(initial: number) {
        const __e = new WithParamsFrameEvent("start", {"initial": initial});
        const __ctx = new WithParamsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public add(value: number) {
        const __e = new WithParamsFrameEvent("add", {"value": value});
        const __ctx = new WithParamsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public multiply(a: number, b: number): number {
        const __e = new WithParamsFrameEvent("multiply", {"a": a, "b": b});
        const __ctx = new WithParamsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    public get_total(): number {
        const __e = new WithParamsFrameEvent("get_total", null);
        const __ctx = new WithParamsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    private _state_Running(__e: WithParamsFrameEvent) {
        if (__e._message === "add") {
            const value = __e._parameters?.["value"];
            this.total += value;
            console.log(`Added ${value}, total is now ${this.total}`);
        } else if (__e._message === "get_total") {
            this._context_stack[this._context_stack.length - 1]._return = this.total;
            return;;
        } else if (__e._message === "multiply") {
            const a = __e._parameters?.["a"];
            const b = __e._parameters?.["b"];
            const result = a * b;
            this.total += result;
            console.log(`Multiplied ${a} * ${b} = ${result}, total is now ${this.total}`);
            this._context_stack[this._context_stack.length - 1]._return = result;
            return;;
        } else if (__e._message === "start") {
            const initial = __e._parameters?.["initial"];
            console.log("Already running");
        }
    }

    private _state_Idle(__e: WithParamsFrameEvent) {
        if (__e._message === "add") {
            const value = __e._parameters?.["value"];
            console.log("Cannot add in Idle state");
        } else if (__e._message === "get_total") {
            this._context_stack[this._context_stack.length - 1]._return = this.total;
            return;;
        } else if (__e._message === "multiply") {
            const a = __e._parameters?.["a"];
            const b = __e._parameters?.["b"];
            this._context_stack[this._context_stack.length - 1]._return = 0;
            return;;
        } else if (__e._message === "start") {
            const initial = __e._parameters?.["initial"];
            this.total = initial;
            console.log(`Started with initial value: ${initial}`);
            const __compartment = new WithParamsCompartment("Running", this.__compartment.copy());
            this.__transition(__compartment);
        }
    }
}


function main() {
    console.log("=== Test 07: Handler Parameters ===");
    const s = new WithParams();

    // Initial total should be 0
    let total = s.get_total();
    if (total !== 0) {
        throw new Error(`Expected initial total=0, got ${total}`);
    }

    // Start with initial value
    s.start(100);
    total = s.get_total();
    if (total !== 100) {
        throw new Error(`Expected total=100, got ${total}`);
    }
    console.log(`After start(100): total = ${total}`);

    // Add value
    s.add(25);
    total = s.get_total();
    if (total !== 125) {
        throw new Error(`Expected total=125, got ${total}`);
    }
    console.log(`After add(25): total = ${total}`);

    // Multiply with two params
    const result = s.multiply(3, 5);
    if (result !== 15) {
        throw new Error(`Expected multiply result=15, got ${result}`);
    }
    total = s.get_total();
    if (total !== 140) {
        throw new Error(`Expected total=140, got ${total}`);
    }
    console.log(`After multiply(3,5): result = ${result}, total = ${total}`);

    console.log("PASS: Handler parameters work correctly");
}

main();

File: output/python/tests/07_params.py
from typing import Any, Optional, List, Dict, Callable

class WithParamsFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class WithParamsFrameContext:
    def __init__(self, event: WithParamsFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class WithParamsCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'WithParamsCompartment':
        c = WithParamsCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class WithParams:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.total = 0
        self.__compartment = WithParamsCompartment("Idle")
        self.__next_compartment = None
        __frame_event = WithParamsFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = WithParamsFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = WithParamsFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = WithParamsFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def start(self, initial: int):
        __e = WithParamsFrameEvent("start", {"initial": initial})
        __ctx = WithParamsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def add(self, value: int):
        __e = WithParamsFrameEvent("add", {"value": value})
        __ctx = WithParamsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def multiply(self, a: int, b: int) -> int:
        __e = WithParamsFrameEvent("multiply", {"a": a, "b": b})
        __ctx = WithParamsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def get_total(self) -> int:
        __e = WithParamsFrameEvent("get_total", None)
        __ctx = WithParamsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def _state_Idle(self, __e):
        if __e._message == "add":
            value = __e._parameters["value"]
            print("Cannot add in Idle state")
        elif __e._message == "get_total":
            self._context_stack[-1]._return = self.total
            return
        elif __e._message == "multiply":
            a = __e._parameters["a"]
            b = __e._parameters["b"]
            self._context_stack[-1]._return = 0
            return
        elif __e._message == "start":
            initial = __e._parameters["initial"]
            self.total = initial
            print(f"Started with initial value: {initial}")
            __compartment = WithParamsCompartment("Running", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)

    def _state_Running(self, __e):
        if __e._message == "add":
            value = __e._parameters["value"]
            self.total += value
            print(f"Added {value}, total is now {self.total}")
        elif __e._message == "get_total":
            self._context_stack[-1]._return = self.total
            return
        elif __e._message == "multiply":
            a = __e._parameters["a"]
            b = __e._parameters["b"]
            result = a * b
            self.total += result
            print(f"Multiplied {a} * {b} = {result}, total is now {self.total}")
            self._context_stack[-1]._return = result
            return
        elif __e._message == "start":
            initial = __e._parameters["initial"]
            print("Already running")


def main():
    print("=== Test 07: Handler Parameters ===")
    s = WithParams()

    # Initial total should be 0
    total = s.get_total()
    assert total == 0, f"Expected initial total=0, got {total}"

    # Start with initial value
    s.start(100)
    total = s.get_total()
    assert total == 100, f"Expected total=100, got {total}"
    print(f"After start(100): total = {total}")

    # Add value
    s.add(25)
    total = s.get_total()
    assert total == 125, f"Expected total=125, got {total}"
    print(f"After add(25): total = {total}")

    # Multiply with two params
    result = s.multiply(3, 5)
    assert result == 15, f"Expected multiply result=15, got {result}"
    total = s.get_total()
    assert total == 140, f"Expected total=140, got {total}"
    print(f"After multiply(3,5): result = {result}, total = {total}")

    print("PASS: Handler parameters work correctly")

if __name__ == '__main__':
    main()
